### ISSUE TITLE
Startup would sometimes exit with error code 2

### ISSUE THREAD

**Issue by whentojump** [whentojump/acto-cloudlab#4]
## TL;DR

We suspect it's an issue with CloudLab.

An easy workaround for the time being is to rerun the startup manually:

```shell
sudo su - geniuser
bash /local/repository/scripts/cloudlab_startup_run_by_geniuser.sh
exit
```

### Behavior

The startup occasionally fails. The "Startup" column will finally become `Exited (2)` instead of `Finished`.

### Possible causes

One of our captured logs says:

```text
TASK [Install python packages using pip] ***************************************
fatal: [127.0.0.1]: FAILED! => {"changed": false, "cmd": ["/usr/bin/python3", "-m", "pip.__main__", "install", "-r", "/users/alice/workdir/acto/requirements.txt"], "msg": "stdout: Collecting deepdiff~=6.3.0\n\n:stderr:  WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f0ca2a1d130>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /packages/fe/b3/81bb598d24f1a48eaceb32243a91016385c0599196a59eaff6cd29299334/deepdiff-6.3.1-py3-none-any.whl\n WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f0ca2a1d0d0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /packages/fe/b3/81bb598d24f1a48eaceb32243a91016385c0599196a59eaff6cd29299334/deepdiff-6.3.1-py3-none-any.whl\n WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f0ca2a1d1f0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /packages/fe/b3/81bb598d24f1a48eaceb32243a91016385c0599196a59eaff6cd29299334/deepdiff-6.3.1-py3-none-any.whl\n WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f0ca2a1d3a0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /packages/fe/b3/81bb598d24f1a48eaceb32243a91016385c0599196a59eaff6cd29299334/deepdiff-6.3.1-py3-none-any.whl\n WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f0ca2a1d5e0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /packages/fe/b3/81bb598d24f1a48eaceb32243a91016385c0599196a59eaff6cd29299334/deepdiff-6.3.1-py3-none-any.whl\nERROR: Could not install packages due to an EnvironmentError: HTTPSConnectionPool(host='[files.pythonhosted.org](http://files.pythonhosted.org/)', port=443): Max retries exceeded with url: /packages/fe/b3/81bb598d24f1a48eaceb32243a91016385c0599196a59eaff6cd29299334/deepdiff-6.3.1-py3-none-any.whl (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f0ca2a1d7c0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))\n\n"}
```

which suggests it's possibly a DNS problem.

The default `resolv.conf` is 

```text
nameserver 130.127.132.51
search clemson.cloudlab.us
nameserver 155.98.60.2
```

of which the entries are within the clemson cluster.

...

### To-do's

Things yet to understand:

- The exact reason. Is it truly DNS?
- Does it have anything to do with machine type, say `c8220`?
- Does it always occur during startup? What about in the middle of experiment?
- ...

What makes things hard is the problem is occasional and unpredictable, thus hard to reproduce.

Long-term solutions:

- Retry. Implemented in #3 
- If we can confirm DNS is the very problem, we may also add some public name servers to `resolv.conf`.
